Search + JSON quick start (#2080)
RedisLabs/redislabs-docs: content/modules/redisearch/search-json-quickstart.md

---
Title: Search JSON quick start
linkTitle: Quick start (with JSON)
description: RediSearch quick start on how to index JSON documents and run search queries.
weight: 22
alwaysopen: false
categories: ["Modules"]
module: RediSearch
aliases: /modules/redisearch/search-json-quickstart/
---

> This quick start shows you how to index documents stored as JSON and run search queries against the index.

Reviewer: ```suggestion
This quick start shows you how to index documents stored as JSON files and run search queries against the index.
```

Reviewer: Or you could say "...index documents formatted with JSON"

## Prerequisites

For this quick start tutorial, you need:

- Either:

    - A [Redis Cloud]({{<relref "/modules/modules-quickstart.md">}}) database with [Redis Stack]({{<relref "/modules/redis-stack">}})

    - A [Redis Enterprise Software]({{<relref "/modules/install/add-module-to-database">}}) database with RediSearch (v2.2 or later) and RedisJSON (v2.0 or later)

> 

Reviewer: Add some indicator here that the two below are not part of the "either" list. Something like AND?

- [`redis-cli`]({{<relref "/rs/references/cli-utilities/redis-cli">}}) command-line tool

- [`redis-py`](https://github.com/redis/redis-py) client library v4.0.0 or later

## Search JSON with `redis-cli`

> The [`redis-cli`]({{<relref "/rs/references/cli-utilities/redis-cli">}}) command-line tool comes packaged with Redis. You can use it to connect to your Redis database and try out RediSearch with RedisJSON.

Reviewer: I would add the description info of redis-cli to first use in the prereqs

To begin, [connect to your database]({{<relref "/rs/references/cli-utilities/redis-cli#connect-to-a-database">}}) with `redis-cli`.

### Create an index

To create an index for JSON documents with [`FT.CREATE`](https://redis.io/commands/ft.create), use the <nobr>`ON JSON`</nobr> keyword.

You also need to include a schema definition that indicates which JSON elements to index. When you define the schema, use a [JSON path]({{<relref "/modules/redisjson#redisjson-paths">}}) expression to map a specific JSON element to a schema field:

```sql
SCHEMA <JSONPath> AS <field_name> <TYPE>
```

The examples in this tutorial use the extended [JSONPath syntax]({{<relref "/modules/redisjson#jsonpath-syntax">}}).

The schema for this example includes four fields: 
- `name` (`TEXT`)
- `description` (`TEXT`)
- `connectionType` (`TEXT`)
- `price` (`NUMERIC`)

This example defines the schema and creates the index:

```sh
127.0.0.1:12543> FT.CREATE itemIdx ON JSON PREFIX 1 item: SCHEMA $.name AS name TEXT $.description as description TEXT $.connection.type AS connectionType TEXT $.price AS price NUMERIC
```

> After you create the index, it automatically adds all existing and future JSON documents prefixed with `item:` to the index.

Reviewer: What is "it"? The schema?

```

#### Return specific fields

You can use [field projection](https://redis.io/docs/stack/search/indexing_json/#field-projection) to limit the search results to include only specific parts of a JSON document. Use the `RETURN` option to specify which fields to include in the search results.

The following query uses the JSONPath expression `$.stock` to return each item's stock in addition to the name and price:

```sh
127.0.0.1:6379> FT.SEARCH itemIdx '@description:(headphones)' RETURN 5 name price $.stock AS stock
1) "2"
2) "item:1"
3) 1) "name"
   2) "Noise-cancelling Bluetooth headphones"
   3) "price"
   4) "99.98"
   5) "stock"
   6) "25"
4) "item:2"
5) 1) "name"
   2) "Wireless earbuds"
   3) "price"
   4) "64.99"
   5) "stock"
   6) "17"
```
### Aggregate search results

The [`FT.AGGREGATE`](https://redis.io/commands/ft.aggregate) command lets you run a search query and modify the results with operations such as `SORTBY`, `REDUCE`, `LIMIT`, `FILTER`, and more. For a detailed list of available operations and examples, see [Aggregations](https://redis.io/docs/stack/search/reference/aggregations).

> You can use [JSON path]({{<relref "/modules/redisjson#redisjson-paths">}}) expressions with the `LOAD` option to run aggregations on JSON documents. You can use any JSON element, even elements that are not included in the index schema.

Reviewer: Just an editorial note: "you can use" is repeated pretty often. Try to switch up the sentence structure occasionally to avoid repetition. i.e. switch between "use x to do y" and "to do y, use x". For me it usually depends whats more important, the action of doing y, or the method of using x is more important.

This example uses aggregation operations to calculate a 10% price discount for each item and then sorts the items from least expensive to most expensive:

```sh
127.0.0.1:6379> FT.AGGREGATE itemIdx '*' LOAD 4 name $.price AS originalPrice APPLY '@originalPrice - (@originalPrice * 0.10)' AS salePrice SORTBY 2 @salePrice ASC
1) "2"
2) 1) "name"
   2) "Wireless earbuds"
   3) "originalPrice"
   4) "64.99"
   5) "salePrice"
   6) "58.491"
3) 1) "name"
   2) "Noise-cancelling Bluetooth headphones"
   3) "originalPrice"
   4) "99.98"
   5) "salePrice"
   6) "89.982"
```

### Drop the index

You can remove the index without deleting any associated documents with the [`FT.DROPINDEX`](https://redis.io/commands/ft.dropindex) command:

```sh
127.0.0.1:12543> FT.DROPINDEX itemIdx
OK
```

## Search JSON with Python

> If you want to use RediSearch within an application, you can use one of these [client libraries](https://oss.redis.com/redisearch/Clients/).

Reviewer: ```suggestion
If you want to use RediSearch within an application, these [client libraries](https://oss.redis.com/redisearch/Clients/) are available.
```
